### PR TITLE
TDL-16383 Refactor code to class based

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
- Converted dictionary of each stream to a class
- Reshuffled sync logic methods from sync.py to `ActiveCampaign` class in the stream.py
- Verified that there is no difference in the catalog and state file of older dict based code and newer code.
- Passed the integration test case locally.

**Note:**
- Removed integration test case line from `config.yml` as no tests available in this PR. 
- So, First PR [#11 ](https://github.com/singer-io/tap-activecampaign/pull/11) needs to be merged which contains all test cases. 
# Manual QA steps
 - Clone old code and newer code locally.
 - Run discover mode and compare catalog.
 - Run sync mode and verify the output.
 
Note:
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_activecampaign/streams.py [hunk 1]
```diff
@@ -13,640 +20,1029 @@
...
+  'deal_group_users': DealGroupUsers,
+  'ecommerce_order_activities': EcommerceOrderActivities,
+  'email_activities': EmailActivities,
+  'goals': Goals,
+  'site_messages': SiteMessages,
+  'sms': Sms
 }
 
+SUB_STREAMS = {
+    'ecommerce_order_products'
```
Comment (NevilParikh14) [inline tap_activecampaign/streams.py:1034]: @prijendev can't we use "parent": "child" format here for the ease to know the parent stream of the mentioned sub-stream?

Comment (prijendev) [reply tap_activecampaign/streams.py]: Updated.

File: .circleci/config.yml [hunk 2]
```diff
@@ -12,25 +12,25 @@ jobs:
             source /usr/local/share/virtualenvs/tap-activecampaign/bin/activate
             pip install -U 'pip<19.2' setuptools
             pip install .[dev]
-      - add_ssh_keys
       - run:
-          name: 'Integration Tests'
+          name: 'JSON Validator'
```
Comment (karanpanchal-crest) [inline .circleci/config.yml:16]: Add comment of code change

File: tap_activecampaign/streams.py [hunk 3]
```diff
@@ -1,3 +1,10 @@
+import singer
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:1]: @prijendev Please add detailed comments in all the functions

File: tap_activecampaign/streams.py [hunk 4]
```diff
@@ -13,640 +20,1029 @@
...
+            singer.write_schema(stream_name, schema, stream.key_properties)
+        except OSError as err:
+            LOGGER.error('OS Error writing schema for: {}'.format(stream_name))
+            raise err
+        
+    def write_record(self, stream_name, record, time_extracted):
+        try:
+            singer.messages.write_record(stream_name, record, time_extracted=time_extracted)
+        except OSError as err:
+            LOGGER.error('OS Error writing record for: {}'.format(stream_name))
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:58]:
```suggestion
            LOGGER.error('OS Error while writing record for: {}'.format(stream_name))
```

File: tap_activecampaign/streams.py [hunk 5]
```diff
@@ -13,640 +20,1029 @@
...
+            LOGGER.error('Stream: {}, record: {}'.format(stream_name, record))
+            raise err
+        except TypeError as err:
+            LOGGER.error('Type Error writing record for: {}'.format(stream_name))
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:62]:
```suggestion
            LOGGER.error('Type Error while writing record for: {}'.format(stream_name))
```

File: tap_activecampaign/streams.py [hunk 6]
```diff
@@ -13,640 +20,1029 @@
...
+                        if bookmark_dttm:
+                            if bookmark_dttm > last_dttm:
+                                self.write_record(stream_name, transformed_record, \
+                                    time_extracted=time_extracted)
+                                counter.increment()
+                    else:
+                        self.write_record(stream_name, transformed_record, time_extracted=time_extracted)
+                        counter.increment()
+
+            return max_bookmark_value, counter.value
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:139]: Add space

**Review by karanpanchal-crest — APPROVED** [review]

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]
Incorporate requested changes in the PR

File: tap_activecampaign/streams.py (hunk 4 above)
Comment (prijendev) [reply tap_activecampaign/streams.py]: Updated

File: tap_activecampaign/streams.py (hunk 5 above)
Comment (prijendev) [reply tap_activecampaign/streams.py]: Updated

File: tap_activecampaign/streams.py (hunk 6 above)
Comment (prijendev) [reply tap_activecampaign/streams.py:160]: Added space

File: tap_activecampaign/streams.py (hunk 3 above)
Comment (prijendev) [reply tap_activecampaign/streams.py:1]: Added comments in all the functions

File: .circleci/config.yml (hunk 2 above)
Comment (prijendev) [reply .circleci/config.yml:16]: Here, `JSON Validator` and `pylint` are added in the `config.yml`.  Removed the integration test case from config.yml as no tests are available in this PR which caused build failures. The existing test case was related to tap-Snapchat. So, we removed that and there are no other test cases available regarding this tap-activecampaign code.

**Review by dbshah1212 — APPROVED** [review]

File: .circleci/config.yml (hunk 2 above)
Comment (karanpanchal-crest) [reply .circleci/config.yml:16]: Also, standard test cases are added as part of PR #11

File: tap_activecampaign/streams.py [hunk 7]
```diff
@@ -13,640 +20,1052 @@
...
+                    if bookmark_field and (bookmark_field in transformed_record):
+                        last_dttm = self.transform_datetime(last_datetime)
+                        bookmark_dttm = self.transform_datetime(transformed_record[bookmark_field])
+                        # Keep only records whose bookmark is after the last_datetime
+                        if bookmark_dttm:
+                            if bookmark_dttm > last_dttm:
+                                self.write_record(stream_name, transformed_record, \
+                                    time_extracted=time_extracted)
+                                counter.increment()
+                    else:
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:155]: @prijendev Please add comments above the `if-else` condition

File: tap_activecampaign/streams.py [hunk 8]
```diff
@@ -13,640 +20,1052 @@
...
+                if isinstance(data, list):
+                    data_list = data
+                    data_dict[data_key] = data_list
+                    transformed_data = transform_json(data_dict, self.stream_name, data_key)
+                elif isinstance(data, dict):
+                    data_list.append(data)
+                    data_dict[data_key] = data_list
+                    transformed_data = transform_json(data_dict, self.stream_name, data_key)
+
+            # LOGGER.info('transformed_data = {}'.format(transformed_data)) # TESTING, comment out
```
Comment (karanpanchal-crest) [inline tap_activecampaign/streams.py:260]: @prijendev please remove this commented code

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

File: tap_activecampaign/streams.py (hunk 7 above)
Comment (prijendev) [reply tap_activecampaign/streams.py:158]: Added comments

File: tap_activecampaign/streams.py (hunk 8 above)
Comment (prijendev) [reply tap_activecampaign/streams.py]: Removed

**Review by karanpanchal-crest — APPROVED** [review]

File: tap_activecampaign/streams.py [hunk 9]
```diff
@@ -13,640 +20,1054 @@
...
+
+    def transform_datetime(self, this_dttm):
+        """
+        Transform the datetime to standard datetime format "%Y-%m%dT%H:%M:%S.000000Z"
+        """
+        with Transformer() as transformer:
+            new_dttm = transformer._transform_datetime(this_dttm)
+        return new_dttm
+
+    def process_records(self,
```
Comment (KrisPersonal) [inline tap_activecampaign/streams.py:108]: Add some comments for process_records. What is this function is expected to do?

Comment (prijendev) [reply tap_activecampaign/streams.py:108]: Added comments in the code.

**Review by KrisPersonal — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]
Approved, but I think the base suite should get merged prior to this PR.

**Pull request merged**